## Introduction

We are going to be using [scikit learn](http://scikit-learn.org/stable/index.html)

<img src='files/resources/scikit-learn-logo-small.png' align='left'><h2>Machine Learning in Python</h2>

<br>
* Simple and efficient tools for data mining and data analysis
* Accessible to everybody, and reusable in various contexts
* Built on NumPy, SciPy, and matplotlib
* Open source, commercially usable - BSD license


In [1]:
import os 
import pandas as pd

In [2]:
data = pd.read_csv("data/fullDataset.tsv", sep="\t",header=0)


/home/csumb/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2902: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
data.head()

key                   query  \
0  215248    all star ticket 2015   
1   42799            mga top bows   
2   73041  cross document markers   
3  138766      wall cabinet white   
4  222660   2.5 jato cooling head   

                                               Title LeafCats  ItemID  \
0  6N/7D Disney World All Star Music~ All Inclusi...    29579   67884   
1  MGTC/TD/TF, MGA, Morris Minor Top Wing Bolts s...    34206   21331   
2  New Cross Vice Gel Ink Pen Gift Set & Document...   165631   36048   
3  White PAPER TOWEL ROLL HOLDER Cabinet Wall Mou...    20643   49003   
4  ENGINE COOLING HEAD TRAXXAS JATO TRX 2.5 NITRO...    34061   71436   

   X_unit_id     SCORE  label_relevanceGrade  label_relevanceBinary  \
0  793072069   -1:1:-1                     3                      0   
1  687419458      3::2                     5                      0   
2  687423583     3:3:3                     6                      1   
3  780091360  -1:-2:-2                     1                      0   
4  793073501     3:3:3                     6                      1   

     feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0  2676.618000          0  -9.020574  -5.221152  32.879559   0.000000   
1  2676.618000          0  -7.435843  -5.986507  24.565182   0.036317   
2  2676.618000          0  -8.030511  -4.509118  48.652847   0.000000   
3  1557.416626          1  -8.218757  -6.132952  45.209755   0.000000   
4  2676.618000          0  -8.599997  -5.317750  79.376099   0.805351   

   feature_7  feature_8  feature_9  feature_10  
0        241   -1000000        139 -100.000000  
1          0          0        208 -100.000000  
2          0          0        226 -100.000000  
3       1258   -1000000        158    1.670978  
4          0          0        150    2.538091

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 78500 entries, 0 to 78499
Data columns (total 19 columns):
key                      78500 non-null int64
query                    78500 non-null object
Title                    78494 non-null object
LeafCats                 78500 non-null object
ItemID                   78500 non-null int64
X_unit_id                78500 non-null object
SCORE                    78500 non-null object
label_relevanceGrade     78500 non-null int64
label_relevanceBinary    78500 non-null float64
feature_1                78500 non-null float64
feature_2                78500 non-null float64
feature_3                78500 non-null float64
feature_4                78500 non-null float64
feature_5                78500 non-null float64
feature_6                78500 non-null float64
feature_7                78500 non-null float64
feature_8                78500 non-null float64
feature_9                78500 non-null int64
feature_10               78499 non-null 

In [5]:
data.describe()

key        ItemID  label_relevanceGrade  \
count   78500.000000  7.850000e+04          78500.000000   
mean    98346.614064  4.869803e+04              5.254777   
std     70253.651231  2.890059e+06              1.334406   
min         1.000000  1.000000e+00              0.000000   
25%     39209.750000  1.908475e+04              5.000000   
50%     78577.500000  3.831350e+04              6.000000   
75%    157546.250000  5.765925e+04              6.000000   
max    234520.000000  8.097468e+08              6.000000   

       label_relevanceBinary     feature_1     feature_2     feature_3  \
count           78500.000000  78500.000000  78500.000000  78500.000000   
mean                0.713653   2821.519304      0.562437     -7.754530   
std                 8.446960    834.067146      0.497410      1.162798   
min                 0.000000  -1890.000000     -9.658937    -13.563713   
25%                 0.000000   2676.618000      0.000000     -8.285197   
50%                 1.000000   2676.618000      1.000000     -7.651813   
75%                 1.000000   3237.216614      1.000000     -7.390667   
max              2363.750000   5286.000000      1.000000     -2.726150   

          feature_4     feature_5     feature_6       feature_7  \
count  78500.000000  78500.000000  78500.000000    78500.000000   
mean      -4.280808     47.526718      0.439879    50746.223210   
std        0.772099     24.222248     18.439044   228258.768038   
min      -10.324562      0.000000      0.000000 -1000000.000000   
25%       -4.736623     30.806008      0.101174        0.000000   
50%       -4.327543     44.189650      0.365236      191.000000   
75%       -3.766841     60.076828      0.613038     4707.250000   
max       42.595760    273.239624   5166.000000  3339990.000000   

            feature_8     feature_9    feature_10  
count    78500.000000  78500.000000  78499.000000  
mean   -240684.450178    171.255248    -62.985598  
std     527649.016958     41.412263     48.710378  
min   -1000000.000000   -100.000000   -100.000000  
25%   -1000000.000000    144.000000   -100.000000  
50%          0.000000    172.000000   -100.000000  
75%      98130.000000    203.000000      0.620428  
max    1000000.000000    240.000000     10.940677

In [6]:
data = data.dropna()
data.describe()

key        ItemID  label_relevanceGrade  \
count   78493.000000  78493.000000          78493.000000   
mean    98353.149287  38385.733390              5.254774   
std     70253.061135  22259.156287              1.334363   
min         1.000000      1.000000              0.000000   
25%     39212.000000  19088.000000              5.000000   
50%     78584.000000  38316.000000              6.000000   
75%    157548.000000  57660.000000              6.000000   
max    234520.000000  77080.000000              6.000000   

       label_relevanceBinary     feature_1     feature_2     feature_3  \
count           78493.000000  78493.000000  78493.000000  78493.000000   
mean                0.683526   2821.567172      0.562598     -7.754471   
std                 0.465103    834.042432      0.496069      1.162636   
min                 0.000000  -1890.000000      0.000000    -13.563713   
25%                 0.000000   2676.618000      0.000000     -8.285108   
50%                 1.000000   2676.618000      1.000000     -7.651790   
75%                 1.000000   3237.285645      1.000000     -7.390667   
max                 1.000000   5286.000000      1.000000     -2.726150   

          feature_4     feature_5     feature_6       feature_7  \
count  78493.000000  78493.000000  78493.000000    78493.000000   
mean      -4.281164     47.528829      0.374066    50762.626642   
std        0.753205     24.222072      0.280593   228237.868119   
min      -10.324562      0.000000      0.000000        0.000000   
25%       -4.736623     30.808287      0.101141        0.000000   
50%       -4.327510     44.190533      0.365175      191.000000   
75%       -3.766839     60.078297      0.613038     4706.000000   
max       -1.106641    273.239624      1.000000  3339990.000000   

           feature_8     feature_9    feature_10  
count    78493.00000  78493.000000  78493.000000  
mean   -240647.46213    171.256023    -62.984023  
std     527631.90141     41.400772     48.710849  
min   -1000000.00000      0.000000   -100.000000  
25%   -1000000.00000    144.000000   -100.000000  
50%          0.00000    172.000000   -100.000000  
75%      98130.00000    203.000000      0.620932  
max    1000000.00000    240.000000     10.940677

In [7]:
from bokeh.charts import defaults, Bar, output_notebook, show
defaults.plot_width = 500
defaults.plot_height = 300
output_notebook(hide_banner=True)

We can see that low relevance results are under represented:

In [8]:
p = Bar(data, 'label_relevanceGrade',  values='label_relevanceGrade', agg='count', title="Distribution of Relevance Grade")
show(p)

For machine learning a lazy algorithm could predict grade 6 for everything.  Predicting the rarer class 0 or 1 has less payoff.  To 'encourage' our algorithm to more discriminating classification we can turn the 6-level relevance score into a binary target.  Now it does not pay to classify everything as relevant.

In [9]:
p = Bar(data, 'label_relevanceBinary',  values='label_relevanceBinary', agg='count', title="Distribution of Relevance Grade")
show(p)

Let's look at the relationship between this binary relevance target and the features.

In [10]:
!pip install seaborn

You are using pip version 8.0.2, however version 8.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [11]:
# %matplotlib inline
# import seaborn as sns;
# sns.set(style="ticks", color_codes=True)
# sns.pairplot(data[["feature_1", "feature_2", "feature_3", "feature_4", "feature_5", "feature_6",
#                   "feature_7", "feature_8", "feature_9", "feature_10"]])

In [17]:
# boxplot feature_1 by relevance label
from bokeh.io import gridplot
from bokeh.charts import BoxPlot, output_file, show

defaults.plot_width = 400
defaults.plot_height = 300

def bplot(feature):
    return BoxPlot(data, values=feature, label="label_relevanceBinary", outliers=False)

p = gridplot ( [ 
        [ bplot("feature_1"), bplot("feature_2") ],
        [ bplot("feature_3"), bplot("feature_4") ],
        [ bplot("feature_5"), bplot("feature_6") ],
        [ bplot("feature_7"), bplot("feature_8") ],
        [ bplot("feature_9"), bplot("feature_10") ] ] ) 
show(p)

## Training / Test data

When machine learning we train on one data set and measure performance on another - this ensures that our performance metrics are representative of the actual performance of the model - this helps avoid overfitting.

In [34]:
from sklearn.cross_validation import train_test_split

X = data[["feature_1", "feature_2", "feature_3", "feature_4", "feature_5", "feature_6",
          "feature_7", "feature_8", "feature_9", "feature_10"]]
y = data[["label_relevanceBinary"]]

X_train, y_train, X_test, y_test = train_test_split(X, y, test_size = 0.25, random_state=73)

print("X training : " + str(X_train.shape))
print("X test     : " + str(y_train.shape))
print("y training : " + str(X_test.shape))
print("y test     : " + str(y_test.shape))

X training : (58869, 10)
X test     : (19624, 10)
y training : (58869, 1)
y test     : (19624, 1)


## Naive Model

Model 01 - just predict everything to be relevant and check performance - this is our baseline.

We now need to define the metric we are going to use to evaluate and compare models.  There are many classification metrics - based on the cells in the confusion matrix - expand this.
 
|                  || *Predicted*  | *Class*   ||
|------------------||--------------------------|||
|                  ||     -    |    +    | Total |
|*True*   |   -     |   TN     |    FP   |  N    |
|*Classs* |   +     |   FN     |    TP   |  P    |
|         | Total   |   N*     |    P*   |       |


Now we can define metrics based on the contingency table:

| Name                      |   Definition  |   Synonyms                                  |
|---------------------------|---------------|---------------------------------------------|
| False Positive Rate       |   FP / N      | Type-I error, 1-Specificity                 |
| True Positive Rate        |   TP / P      | 1-Type II error, power, sensitivity, recall |
| Positive Predictive Value |   TP / P*     | Precision, 1-false discovery proportion     |
| Negative Predictive Value |   TN / N*     |                                             |


Visual for this at https://en.wikipedia.org/wiki/Precision_and_recall - be careful not to introduce confusion here.  For the course we could just optimise on F1?

Sometimes we talk about the accuracy of a model as:

$
accuracy = \frac{(TP + TN)}{( N + P )}
$


Generally we are interested in precision and recall.  

* Precision
    * The fraction of cases predicted to be positive that are actually positive
    * High precision is acheived by ensuring that all the predicted positive cases are actually positive
    * Quality
* Recall
    * The fraction of the positive cases predicted to be positive 
    * High recall is acheived by a model predicts as many of the positive cases as possible 
    * Quantity

We can also define classification metrics that compbine both precision and recall.  The F1-score is one such metric that computes the geometric mean of precision and recall - applying equal emphasis to precision as recall:

$
F_1 = 2 \times \frac{ precision . recall}{precision + recall} 
$

But we may choose to attach emphasis to recall:

$
F_2 = 5 \times \frac{ precision . recall}{ (4 . precision) + recall} 
$

Or we may choose to place higher emphasis on precision:

$
F_{0.5} = 1.25 \times \frac{ precision . recall}{( 0.25 . precision) + recall} 
$



In [47]:
y_pred_01 = [1] * y_test.shape[0]

from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
print(confusion_matrix(y_test, y_pred_01))
print(classification_report(y_test, y_pred_01))
print(accuracy_score(y_test, y_pred_01))

[[    0  6198]
 [    0 13426]]
             precision    recall  f1-score   support

        0.0       0.00      0.00      0.00      6198
        1.0       0.68      1.00      0.81     13426

avg / total       0.47      0.68      0.56     19624

0.684162250306


We can see that recall is 100% - we predicted every value that was positive correctly - but precision is low - of those predicted to be positve only 68% were actually postive.  This yields and F1 score of 81%.  This is our baseline.

## Support Vector Machine

SVMs are the default go to classification algorithm.  They are fast, scalable and relatively simple despite the scary name!

To fit an SVM, and in general when performing Machine Learning, we scale the data to have zero mean and unit standard deviation - this ensures that the algorithms do not over rely on features with large values.


